### PR TITLE
Don't send service LoadBalancer traffic through master nodes

### PR DESCRIPTION
**Pull request by justinsb**
We are moving to mark masters as tainted, and not necessarily
unschedulable.  But this means we are now sending general service
traffic through the master.

Recognize "well-known" master taints and labels, and avoid attaching
these nodes to LoadBalancers.

Issue #33884

(Does not completely resolve the issue, because we also should not even
be opening the NodePort)

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35902)

### CODE REVIEW

**Comment by mikedanese** [comment]
@k8s-bot unit test this

**Comment by mikedanese** [comment]
> But this means we are now sending general service
> traffic through the master.

What health check are we using for elbs?

File: pkg/controller/service/servicecontroller.go [hunk 1]
```diff
@@ -593,14 +593,57 @@ func hostsFromNodeSlice(nodes []*api.Node) []string {
...
+	if role := node.Labels["kubeadm.alpha.kubernetes.io/role"]; role != "" {
+		return role
+	}
+
+	taints, err := api.GetTaintsFromNodeAnnotations(node.Annotations)
+	if err != nil {
+		glog.Warningf("error parsing node taints for node %q: %v", node.Name, err)
+	} else {
+		for _, taint := range taints {
+			if taint.Key == "dedicated" {
```
Comment (mikedanese) [inline pkg/controller/service/servicecontroller.go]: We should figure out what we are going to call this, ref https://github.com/kubernetes/kubernetes/pull/31647#issuecomment-2568505

**Comment by mikedanese** [comment]
cc @davidopp

**Comment by justinsb** [comment]
> What health check are we using for elbs?

TCP level health checks on the NodePort.  Where we have a service with multiple ports, we check the first port.  If we do ESIPP that may get more complex.

**Comment by justinsb** [comment]
So let's figure out how to fix this:

* we could trigger on the 'kubernetes.io/role' family of labels (cf #35975), and not the taints - apparently our usage of them is not as intended.
* we could also invent a new label `kubernetes.io/disable-node-port` or similar

Do people like those options or have another suggestion?  cc @thockin @mikedanese @caseydavenport

**Comment by caseydavenport** [comment]
@justinsb My preference is not to re-use the `kubernetes.io/role` labels.  Masters can also run pods sometimes (so you might want nodeports).

A dedicated label would work for me.

**Comment by dims** [comment]
@mikedanese @justinsb : will this make it to v1.5? if not, can you please set the milestone to v1.6

**Comment by justinsb** [comment]
@dims it is a bug, so yes we should fix it in 1.5, or at least have a workaround available.

@thockin do you agree that a dedicated label is best (vs reusing the role label)

@caseydavenport any ideas as to what label to use?

How about k8s.io/nodeport=no on the master node?

**Comment by caseydavenport** [comment]
>  will this make it to v1.5?

I don't think needs to go into v1.5 - feature freeze has occurred, has it not? IIUC, impact of this is that traffic goes through the master, which might be undesirable for perf reasons but shouldn't cause things to fail?

> How about k8s.io/nodeport=no on the master node?

What about `net.alpha.kubernetes.io/disableNodePort=true|false` ?

**Comment by mikedanese** [comment]
@davidopp and @dchen1107 say we shouldn't overload dedicated.

**Comment by thockin** [comment]
If the master is running kube-proxy, why NOT run traffic through it?  If we de-centralize more, there may not be "a master" - what then?

The real question, I think is whether a node expects to be participating in the service "mesh" (hate the term, but oh well).  If the master is not running proxy, won't healthchecks fail anyway, so it should not get traffic.

@bprashanth

**Comment by dims** [comment]
@justinsb @thockin @mikedanese : so is this too late for 1.5? (move it to 1.6 please)

**Comment by mikedanese** [comment]
We at least need to remove the section looking for a taint. 

> TCP level health checks on the NodePort. Where we have a service with multiple ports, we check the first port

It sounds like this won't send traffic to a master unless it's running kube-proxy in which case why not?

**Comment by thockin** [comment]
I'm not sure it won't send traffic to a master without kube-proxy - I'd
have to convince myself of that.  Justin?

**Comment by justinsb** [comment]
We run kube-proxy on the master as well (by default) with kops, so that daemonset pods on the master can't reach pods.  The problem is the performance/security implications of opening ports and putting additional load on the master.

**Comment by thockin** [comment]
I am torn - the concept of "the master" is slowly dissipating as
controllers become plain old pods in so many ways.

I guess I am OK with this overall as an idea - it specifically calls out
"dedicated master" but it's one more wrinkle of complexity in reasoning
about behavior.  In this case I guess it maps to normal expectations.
 @mikedanese ?

**Comment by mikedanese** [comment]
Two problems with the current approach:

1. I can see usecases where I would want kube-proxy to send traffic through master nodes. I can also usecases where I would want to send traffic through unscheduable nodes (which is currently not allowed) but we can phase that behavior out when we phase out unschedulable:  #29178. I don't think it's wise to conflate these behaviors.
1. We shouldn't use "dedicated=master" taint yet. @davidopp doesn't want us to jump the gun on dedicated nodes since the naming and implementation is not designed yet: https://github.com/kubernetes/kubernetes/issues/36272.

Hate to say it but the fastest way to deliver this feature would probably be an annotation.

**Comment by dims** [comment]
@justinsb Move the milestone please if appropriate?

**Comment by davidopp** [comment]
> We shouldn't use "dedicated=master" taint yet. @davidopp doesn't want us to jump the gun on dedicated nodes since the naming and implementation is not designed yet: #36272.

I'm fine with "role=master" taint. I just don't want the key to be "dedicated" because I anticipate we will use that to dedicated nodes to groups of users (i.e. the value will be some identifier of a group of users).

**Comment by davidopp** [comment]
> I'm fine with "role=master" taint. I just don't want the key to be "dedicated" because I anticipate we will use that to dedicated nodes to groups of users (i.e. the value will be some identifier of a group of users).

That said, we should only do this if we think the "out of the box" configuration should be to disallow regular user pods from scheduling on the master, *or* if we make every cluster deployment tool explicitly ask whether users want that behavior and use that to decide whether to set the taint when master node is created.

**Comment by justinsb** [comment]
So it sounds like we need a dedicated "marker" rather than having the master role do double-duty.
 
Instead of a label/annotation, i'm inclined to add a field to the Node, because moving from labels to fields is both inevitable and painful, but we have a well-established process for evolving / defaulting fields. (Or the feature is deprecated, and we just have a procedure for that also.)

**Pull request closed without merging**